### ISSUE TITLE
The request object passed to middleware in PHPUnit tests has no route() method

### ISSUE THREAD

**Issue by kfirba** [laravel/lumen-framework#199]
Hello.

The `$request` variable that is being passed to the middlewares in testing mode (the built in PHPUnit tests) has no `route()` method on it, hence it's breaking my middleware code and makes it impossible to write acceptance tests.

If I try to access the same URI in a browser or in postman, there is a `route()` method, hence the result is desirable. 

I tried to figure the source of the error and it seems like that when I access the `$request` parameter from the PHPUnit tests there is no `routeResolver` being set, which keeps it null, hence the `null` result when trying to do `$request->route()`.

On the other hand, if I peek into the request when I test it manually from browser/postman, the `routeResolver` is:

```php
Closure {#83
    class: "Laravel\Lumen\Application"
    this: Application {#3 …}
    file: "/home/vagrant/code/SomeProject/vendor/laravel/lumen-framework/src/Application.php"
    line: "680 to 682"
  }
```

Also, I guess it worth mentioning that the same is for the `userResolver` where it's being NULL in test environment but it set on browser/postman.

Since Lumen acts differently on testing mode and on "production" mode it's really hard to test it.

Maybe I'm doing something wrong and it's not Lumen?

**Comment by kovagoz** [comment]
Hi @kfirba 

This is because `routeResolver` is set in `Application::registerRequestBindings()` which is never called because `Request` has been already created and injected by `CrawlerTrait::call()` via `TestCase`.


**SOLUTION**

Extend `Laravel\Lumen\Application` and replace original instance in `bootstrap/app.php` with yours. Add this new method to the extended Application:

```php
public function getCurrentRoute()
{
    return $this->currentRoute;
}
```

Create a new CrawlerTrait only with this one method:

```php
public function call($method, $uri, $parameters = [], $cookies = [], $files = [], $server = [], $content = null)
{
    $this->currentUri = $this->prepareUrlForRequest($uri);

    $request = Request::create(
        $this->currentUri, $method, $parameters,
        $cookies, $files, $server, $content
    );

    // These three lines were added compared to the original CrawlerTrait.
    $request->setRouteResolver(function () {
        return $this->app->getCurrentRoute();
    });

    return $this->response = $this->app->prepareResponse($this->app->handle($request));
}
```

And finally use your new trait in TestCase:

```php
class TestCase
{
    use CrawlerTrait;
}
```

It will override the original `call()` method and use `getCurrentRoute()` from the patched Application to set the route resolver on request.

**Comment by HenkPoley** [comment]
Is this still the recommended fix for Laravel Lumen 5.8/6.x ?

Is this documented on https://laravel.com/docs/6.x/testing or https://lumen.laravel.com/docs/6.x/testing ?
